Ensure Angular docsMode flag has effect

alexandertrefz opened the pull request:
## What I did

Fix what #20608 was setting out to do: Fixing docs generation for Angular. 

#20608 fixed the issue that the angular builders accept a `docsMode` flag, but were internally attempting to read a flag called `docs` to pass that along to `runInstance`, thereby always reading `undefined` and passing it along.

#20608 fixed this successfully, but introduced a new bug: It now passes along the `docsMode` flag, reading the value correctly, but `runInstance` doesn't accept this flag, it looks for a flag called `docs`.
It accepts a type called `StandaloneOptions`, which is a combination of `CLIOptions` and some other types, `CLIOptions` has the optional property `docs` as can be verified here: https://github.com/storybookjs/storybook/blob/next/code/lib/types/src/modules/core-common.ts#L159

I now simply read the `docsMode` flag but pass this along as `docs`. This fixes the generation successfully.

alexandertrefz committed:
fix: Ensure Angular docsMode flag has effect & rename it to docs

alexandertrefz commented:
/cc @ndelangen

valentinpalkovic commented:
@tmeasday, @JReinhold You have more knowledge about this. What does the error mean?

https://app.circleci.com/pipelines/github/storybookjs/storybook/39103/workflows/859db4a7-a376-41aa-b174-2651f8234a45/jobs/473122/steps?invite=true#step-104-32

And:
https://app.circleci.com/pipelines/github/storybookjs/storybook/39103/workflows/859db4a7-a376-41aa-b174-2651f8234a45/jobs/473122/steps

JReinhold commented:
Hmm it means that there is no Story Indexer that can handle `*.stories.mdx` files, even though they are in the glob. Maybe we removed the indexer to handle story definitions in MDX, and forgot to remove this? I'm unsure, and I don't know if it is related to this PR or not.

@tmeasday or @shilman would know more.

alexandertrefz commented:
This PR changes 2 lines, in passing along a single argument to reenable docs generation for Angular. It literally just renames a single key in an object (in 2 separate files) - I could not imagine it possibly impacting Story indexing in any way.

sheriffMoose commented:
Hey @valentinpalkovic, @JReinhold 
currently the common preset indexer works with `.js|.jsx|.ts|.tsx` and the indexing for `.mdx` resides in the `addon-docs`

tmeasday commented:
Y'all have this backwards I suspect.

The CLI option is `--docs`. It should set the config option `docs.docsMode = true`.

I think what this is doing is setting `docs = <value of cli flag>`, so when the CLI flag is unset, it is setting `docs = false | null | whatever`, which has the effect of disabling docs. So now `.stories.mdx` files are not being indexed, as that indexer is added by the docs addon.

ssams commented:
> #20608 fixed this successfully, but introduced a new bug: It now passes along the `docsMode` flag, reading the value correctly, but `runInstance` doesn't accept this flag, it looks for a flag called `docs`. It accepts a type called `StandaloneOptions`, which is a combination of `CLIOptions` and some other types, `CLIOptions` has the optional property `docs` as can be verified here: https://github.com/storybookjs/storybook/blob/next/code/lib/types/src/modules/core-common.ts#L159

interesting, when I created the original PR it certainly worked with the latest v7/next at that time (the patch from the linked PR for sure works with current v6 releases, as I'm using that locally). Something must have changed internally since then?
Note that in addition to the `CLIOption` you've linked, which is called indeed just `docs`: https://github.com/storybookjs/storybook/blob/8073d0419b724ff69702401908e36547d2b1290e/code/lib/types/src/modules/core-common.ts#L159

There's also another `docsMode` in the `BuilderOptions` type just a few lines later in the file:
https://github.com/storybookjs/storybook/blob/8073d0419b724ff69702401908e36547d2b1290e/code/lib/types/src/modules/core-common.ts#L170

I'd assume it worked so far as it directly set the latter one and it got forwarded / kept included in the config option until reaching the actual builder, could be that types have been improved or gotten stricter in the meantime and it's only now filtered out somewhere in the call chain.

Also at the end of the Angular builder `buildStaticStandalone` is called with the collected options. This function apparently accepts a combination of `CLIOptions` and `BuilderOptions`, so from the type alone both `docs` or `docsMode` would be valid to pass (but apparently they don't work equally in practice):
https://github.com/storybookjs/storybook/blob/8073d0419b724ff69702401908e36547d2b1290e/code/frameworks/angular/src/builders/build-storybook/index.ts#L125
https://github.com/storybookjs/storybook/blob/8073d0419b724ff69702401908e36547d2b1290e/code/lib/core-server/src/build-static.ts#L37-L41

alexandertrefz commented:
> Y'all have this backwards I suspect.
> 
> The CLI option is `--docs`. It should set the config option `docs.docsMode = true`.
> 
> I think what this is doing is setting `docs = <value of cli flag>`, so when the CLI flag is unset, it is setting `docs = false | null | whatever`, which has the effect of disabling docs. So now `.stories.mdx` files are not being indexed, as that indexer is added by the docs addon.

This was a good hint!

I know for certain that we need to set the `docs` flag for the docsMode to kick in. But you were right that my change does indeed seem to turn off the addon-docs in its entirety when `docsMode` is not set. I was able to reproduce the behaviour locally.

I have updated the PR to only set `docs` when `docsMode` is truthy, which has remedied the situation, at least locally.

alexandertrefz commented:
Some fighting with the linter later, and it all seems to run through green. 🎉 

I guess someone needs to approve the other 2 workflows?

tmeasday commented:
Ok, my apologies there was more going on here than I realised.

So this is pretty confusing. There is:

- [`CLIOptions['docs']`](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/types/src/modules/core-common.ts#L159)
- [`BuilderOptions['docsMode']`](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/types/src/modules/core-common.ts#L170) - which is directly set via the above.

Then there is:

- [`StandaloneOptions = CLIOptions & BuilderOptions`](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/frameworks/angular/src/builders/utils/standalone-options.ts#L4-L6) -- which means it has both fields I guess (and they mean the same thing?)

(The above is angular, but there's also a [core-server version of that](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/core-server/src/build-static.ts#L37-L40))

Additionally, there is:

- [StorybookOptions['docs']](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/types/src/modules/core-common.ts#L380), which is an object, with a key `docsMode`.

I guess my suspicion about the indexer was wrong then? It seems an awful coincidence. Or could it be that [`options.presets.apply('docs')`](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/core-server/src/utils/getStoryIndexGenerator.ts#L30) looks at `options.docs` somehow?

alexandertrefz commented:
> Then there is:
> 
> * [`StandaloneOptions = CLIOptions & BuilderOptions`](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/frameworks/angular/src/builders/utils/standalone-options.ts#L4-L6) -- which means it has both fields I guess (and they mean the same thing?)
>

Actually no! They don't do the same thing. `docsMode` has been set since the original PR from last week (#20608) but it does not turn on `--docs` properly. Only passing `docs` to `StandaloneOptions` does turn on `--docs`. However, passing a falsy value explicitly creates the MDX issues we observed in the beginning of this PR. Leaving `docs` out of the option entirely if it is falsy seems to create the default behaviour for the non-angular CLI.

I am unsure if `docsMode` has any effect at this point, but im not familiar enough with the codebase to make a determination either way.

tmeasday commented:
@alexandertrefz I think the reason for that (needing to set `options.docs = true` to turn on docs mode) is due to this line:

https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/core-server/src/presets/common-preset.ts#L207-L214

The second argument to `docs()` there (a preset property) is really the whole `Options` object that includes both `.docs` and `.docsMode` but we only read `options.docs` when setting the default `'docs.docsMode'` preset property.

That preset property is then read by:
 - The index generator [setup code](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/core-server/src/utils/getStoryIndexGenerator.ts#L30)
 - The various builders ([manager](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/builder-manager/src/utils/data.ts#L17), [webpack](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/builder-webpack5/src/preview/iframe-webpack.config.ts#L88), [vite](https://github.com/storybookjs/storybook/blob/2cceaadef83bcc109af4144030a5e79af1a9d48f/code/lib/builder-vite/src/transform-iframe-html.ts#L12)).

I can't see any other places that either `options.docs` or `options.docsMode` is read 🤷

alexandertrefz commented:
@tmeasday So what should be the course of action? Should we attempt to change the `common-preset` to look at both `docs` and `docsMode` then?

tmeasday commented:
OK, I tracked it down. The problem is the essentials addon drops the docs addon if `PresetOptions['docs']` is false here:

https://github.com/storybookjs/storybook/blob/1c7d2fa952428980a7b563a36ba8a173d1b7fe3f/code/addons/essentials/src/index.ts#L51

So what's happening is that when we set `options.docs = false | undefined` it triggers the above. Setting it true works fine (thus the current `docsMode ? { docs: docsMode } : {}`).

This simulates what SB proper does where the CLI option is left unset if you don't pass `--docs` (so you don't end up with `options.docs = false`).

Setting `options.docsMode` does nothing I'm pretty sure (right now) and probably would just confuse things on this PR.

However, this is all a big mess. @shilman we should discuss this.

tmeasday commented:
Ok, discussed with @shilman.

It's pretty weird that we create that mess of `StandaloneOptions` (which can include both `docs` and `docsMode` booleans) and then combine it with the preset's actual options (in this case the options of the essential addon, which has a key `docs`). However, it would be a breaking change to do anything about the latter, and we are kind of stuck on the former as the CLI flag is called `--docs` also and we don't want to change it.

So the TLDR is .. it's super confusing, but the main thing is we do not ever want to set `StandaloneOptions['docs']` to `false`. We should probably drop the unused `docsMode` boolean because it is just muddying the waters (and doesn't do anything AFAICT), but it doesn't need to happen in this PR. You could drop it from the angular side if you wanted.

So we can merge this PR as is or drop the `docsMode,` line.

ndelangen reviewed (approved):
@alexandertrefz great work, could you apply the changes @tmeasday is asking about? then it's good to merge!

alexandertrefz commented:
@tmeasday should we then also rename the angular flag to align with the rest and name it docs? So [here](https://github.com/storybookjs/storybook/blob/next/code/frameworks/angular/src/builders/start-storybook/schema.json#L64) renaming `docsMode` to `docs` and the same [here](https://github.com/storybookjs/storybook/blob/next/code/frameworks/angular/src/builders/build-storybook/schema.json#L37)? 

So that `docsMode` is entirely gone?

alexandertrefz commented:
I have added the requested changes and renamed the `docsMode` flag in the angular builders to `docs` - this should probably also be updated in the documentation for v7 Angular? I have updated a reference I found in the getting started guide, but I am unsure where the proper docs are located for this.

alexandertrefz commented:
@ndelangen Just needs approval for the final workflows now and then it can get merged!

ndelangen commented:
Invited you to the org @alexandertrefz, will merge as soon as CI is green. Awesome work, thank you for contributing!

tmeasday commented:
Thank you for working through this with us!
